### PR TITLE
Fix bug when subset of model parameters is passed into optimizer with FSDP

### PR DESCRIPTION
**Pull request by sashaDoubov**
# What does this PR do?

This PR addresses https://github.com/mosaicml/composer/issues/3493, where we were adding all model parameters into the optimizer state after FSDP wrapping. This change uses the previous flow for param_groups > 1, where we:
1. Get the mapping between params in the optimizer group and their names by using the pointers (found with `id`) of the tensors
2. Store the extra param group info, such as LR, wd, etc. in a separate dict
3. Iterate over the fsdp wrapped parameters, and populates the param_groups based on the names of the fsdp wrapped tensor and the mapping from before.

This change relies on `use_orig_params=True`, so an error is thrown if a subset of parameters is passed into the param_groups.
 
**NOTE**: at the moment, tensor parallelism breaks with multiple param_groups (as the ptrs of TP'd model.parameters() != ptrs of elements in optimizer param_groups), so a special case was added to handle this.

# What issue(s) does this change relate to?
Fixes https://github.com/mosaicml/composer/issues/3493

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/distributed/dist_strategy.py [hunk 1]
```diff
@@ -261,37 +267,41 @@ def sync_hook(*args):
...
                 optimizer_specific_group_info = {
                     k: v for k, v in optim.param_groups[group_num].items() if k != 'params'
                 }
-                group_num_to_param_group_info[group_num] = optimizer_specific_group_info
+                group_num_to_opt_group_info[group_num] = optimizer_specific_group_info
         else:
+            if len(optim.param_groups) > 1:
+                raise RuntimeError(
+                    'Multiple optimizer groups with FSDP are not supported with tensor parallelism and/or use_orig_params=False.',
+                )
```
Comment (sashaDoubov) [inline composer/distributed/dist_strategy.py:298]: would like feedback if we want this error message broken out into separate conditions for TP vs. use_orig_params=False

File: composer/distributed/dist_strategy.py [hunk 2]
```diff
@@ -213,6 +219,7 @@ def prepare_fsdp_module(
         precision: (Precision): The precision being used by the Trainer, used to fill in defaults for FSDP `mixed_precision` settings.
         device (Device): The device being used by the Trainer.
         auto_microbatching (bool, optional): Whether or not auto microbatching is enabled.
+        using_tp (bool, optional): Whether the model has been wrapped with Tensor Parallelism, in which case only a single optimizer param group is supported.
```
Comment (sashaDoubov) [inline composer/distributed/dist_strategy.py:222]: new arg, perhaps there's an easier way to determine if using tp, but this was what made sense to me. would also appreciate feedback regarding this.

File: composer/distributed/dist_strategy.py [hunk 3]
```diff
@@ -171,9 +171,14 @@ def _recreate_fsdp_param_groups_from_unwrapped_opt_info(
     for fsdp_name, param in fsdp_wrapped_named_params:
 
         unwrapped_name = clean_tensor_name(fsdp_name)
-        # need to have a 1:1 mapping between a fsdp param name and the non-wrapped vanilla param name
-        retrieved_group_num = non_wrapped_param_names_to_group_num[unwrapped_name]
-        group_num_to_optimizer_info[retrieved_group_num]['params'].append(param)
+
+        # since we are iterating over all model.named_parameters() after fsdp wrapping, we need to check
```
Comment (mvpatel2000) [inline composer/distributed/dist_strategy.py:175]:
```suggestion
        # Since we are iterating over all model.named_parameters() after fsdp wrapping, we need to check
```
capitalize start of comment

File: composer/distributed/dist_strategy.py [hunk 4]
```diff
@@ -171,9 +171,14 @@ def _recreate_fsdp_param_groups_from_unwrapped_opt_info(
...
+        # if the parameter was included in the optimizer param_group pre fsdp wrapping, in order to support
+        # passing a subset of model params in the optimizer
+        if unwrapped_name in non_wrapped_param_names_to_group_num:
+            # need to have a 1:1 mapping between a fsdp param name and the non-wrapped vanilla param name
```
Comment (mvpatel2000) [inline composer/distributed/dist_strategy.py:179]:
```suggestion
            # Need to have a 1:1 mapping between a fsdp param name and the non-wrapped vanilla param name
```

File: composer/distributed/dist_strategy.py (hunk 2 above)
Comment (mvpatel2000) [reply composer/distributed/dist_strategy.py:222]: can we just check the param groups in `prepare_tp_module`

File: composer/distributed/dist_strategy.py (hunk 1 above)
Comment (sashaDoubov) [reply composer/distributed/dist_strategy.py]: I've now moved this to TP as recommended by @mvpatel2000

File: tests/trainer/test_tp.py [hunk 5]
```diff
@@ -47,3 +49,80 @@ def test_tp_train(world_size: int):
...
+            parallelism_config={
+                'tp': {
+                    'layer_plan': layer_plan,
+                    'tensor_parallel_degree': 2,
+                },
+                'fsdp': {},
+            },
+            max_duration='3ba',
+        )
+    gc.collect()
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:128]: why is this necessary?

File: tests/trainer/test_tp.py [hunk 6]
```diff
@@ -47,3 +49,80 @@ def test_tp_train(world_size: int):
...
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:93]: same as below

**Review by mvpatel2000 — COMMENTED** [review]
LGTM! 

why are gc collect necessary? will approve after that

File: tests/trainer/test_tp.py (hunk 5 above)
Comment (sashaDoubov) [reply tests/trainer/test_tp.py:128]: I was following my previous impl here: https://github.com/mosaicml/composer/blob/98a87f3845aa19d0dee53767c448533ac93287b6/tests/trainer/test_fsdp_param_groups.py#L51, but I'll try removing and see if the test pass in that case.

**Review by mvpatel2000 — APPROVED** [review]
